crates/ruff_linter/src/rules/flake8_comprehensions/rules/unnecessary_map.rs

```diff
@@ -67,140 +64,91 @@ impl Violation for UnnecessaryMap {
-            // Exclude the parent if already matched by other arms.
-            if parent
-                .and_then(Expr::as_call_expr)
-                .and_then(|call| call.func.as_name_expr())
-                .is_some_and(|name| matches!(name.id.as_str(), "list" | "set" | "dict"))
-            {
-                return;
-            }
-
-            // Only flag, e.g., `map(lambda x: x + 1, iterable)`.
-            let [Expr::Lambda(ast::ExprLambda {
-                parameters, body, ..
-            }), iterable] = args
-            else {
-                return;
-            };
-
-            // For example, (x+1 for x in (c:=a)) is invalid syntax
-            // so we can't suggest it.
-            if any_over_expr(iterable, &|expr| expr.is_named_expr()) {
-                return;
-            }
+    let parent = semantic.current_expression_parent();
 
-            if !lambda_has_expected_arity(parameters.as_deref(), body) {
-                return;
-            }
-        }
-        ObjectType::List | ObjectType::Set => {
-            // Only flag, e.g., `list(map(lambda x: x + 1, iterable))`.
-            let [Expr::Call(ast::ExprCall {
-                func,
-                arguments: Arguments { args, keywords, .. },
-                ..
-            })] = args
-            else {
-                return;
+    let (parameters, body, iterables) = match object_type {
+        ObjectType::Generator => {
+            let parent_call_func = match parent {
+                Some(Expr::Call(call)) => Some(&call.func),
+                _ => None,
             };
 
-            if args.len() != 2 {
-                return;
-            }
-
-            if !keywords.is_empty() {
+            // Exclude the parent if already matched by other arms.
+            if parent_call_func.is_some_and(|func| is_list_set_or_dict(func, semantic)) {
                 return;
             }
 
-            let Some(argument) = helpers::first_argument_with_matching_function("map", func, args)
-            else {
+            let Some(result) = map_lambda_and_iterables(call, semantic) else {
                 return;
             };
 
-            let Expr::Lambda(ast::ExprLambda {
-                parameters, body, ..
-            }) = argument
-            else {
-                return;
-            };
-
-            if !lambda_has_expected_arity(parameters.as_deref(), body) {
-                return;
-            }
+            result
         }
-        ObjectType::Dict => {
-            // Only flag, e.g., `dict(map(lambda v: (v, v ** 2), values))`.
-            let [Expr::Call(ast::ExprCall {
-                func,
-                arguments: Arguments { args, keywords, .. },
-                ..
-            })] = args
-            else {
-                return;
-            };
 
-            if args.len() != 2 {
+        ObjectType::List | ObjectType::Set | ObjectType::Dict => {
+            if !arguments.keywords.is_empty() {
                 return;
```

Comment:
This check is now redundant with the check on line 71 or was it unintentional to move the check to line 71 because it wasn't applied for the `Generator` case before